README.md
# WebSphere Cookbook

[![Build Status](https://travis-ci.org/Sainsburys/chef-websphere.svg?branch=master)](https://travis-ci.org/Sainsburys/chef-websphere)

## Scope

This cookbook installs and configures WebSphere Application Server Network Deployment (WAS-ND). It has only been tested with a Deployment Manager and Managed nodes.

So far it has only been tested with WebSphere ND 7.0 and 8.5.

The cookbook lacks idempotency as querying WAS for the current state is overly complex, really slow, or just not possible.  For that reason it's better to destroy and recreate objects rather than update them at this stage.

## Requirements

* Chef 12.5 or higher
* Installation media for Installation manager either local or via a url, and Media or repositories for any other packages like WAS ND, IHS etc.
* You will first need to install WebSphere Network Deployment version using the ibm-installmgr cookbook.  See `test/fixtures/cookbooks/websphere-test/recipes/was_install.rb` for an example.

## Platform

* CentOS 6
* Red Hat Enterprise 6

## Versions

• WASND 7.0
• WASND 8.5

## Usage

The general pattern for usage is to:

1. Instal Installation Manager with the installation manager cookbook
2. Install WebSphere ND, WebSphere Toolbox and WebSphere Plugins (for IHS)
3. Deploy a Deployment Manager
4. Deploy a custom profile
5. Deploy an appserver OR Deploy a cluster and add a cluster member to the profile
6. Deploy IHS if needed.

### Install

You first need to install the necessary components. To do this use the ibm-installmgr cookbook.  It can be used to install any of the following:

* WebSphere ND
* Any WebSphere ND Fixpacks
* IBM Java version(s)
* WebSphere Plugins (PLG)
* WebSphere customization toolbox
* IHS HTTP Server.

As well as the below example, see `test/fixtures/cookbooks/websphere-test/recipes/was_install.rb` for a more complete example.

#### Example

  ```ruby
  install_mgr 'ibm-im install' do
    install_package 'https://someurl/agent.installer.linux.gtk.x86_64_1.8.4000.20151125_0201.zip'
    install_package_sha256 '28f5279abc28695c0b99ae0c3fdee26bfec131186f2ca7e41d1317e303adb12e'
    package_name 'com.ibm.cic.agent'
    ibm_root_dir '/opt/IBM'
    service_user 'ibm-im'
    service_group 'ibm-im'
  end

  ibm_package 'WAS ND install' do
    package 'com.ibm.websphere.ND.v85'
    install_dir '/opt/IBM/WebSphere/AppServer'
    repositories ['/my/path/to/local/install/media']
    action :install
  end
  ```

### websphere_base

Do not use this directly it can be ignored. This resource is only used as a parent class to share methods between child resources.

### websphere_dmgr

#### Example

```ruby
websphere_dmgr 'Dmgr01' do
  cell_name 'MyNewCell'
  admin_user 'admin'
  admin_password 'mypassword'
  action [:create, :start]
end
```

#### Parameters

* `label`, String, name_property: true, just a label to make the execute resource name unique
* `websphere_root`, String, default: '/opt/IBM/WebSphere/AppServer', The was install root
* `bin_dir`, String, default: lazy { "#{websphere_root}/bin" } The path to the was root bin directory. No need to change this.
* `admin_user`, [String, nil], default: nil, The dmgr user.
* `admin_password`, [String, nil], default: nil, The dmgr password.  This should only be known to people who know the server is managed with chef.
* `dmgr_host`, String, default: 'localhost', The dmgr host/ip.
* `dmgr_port`, [String, nil], default: nil, The dmgr port, when nil WAS will default to dmgr SOAP connector default port 8879.
* `profile_name`, String, name_property: true. # The Dmgr profile name.
* `profile_env`, [Hash, nil], default: nil, additional environment variables to send to the start command
* `node_name`, String, default: lazy { "#{profile_name}_node" }. It's easier to keep the default name here.
* `cell_name`, [String, nil], default: nil
* `java_sdk`, [String, nil], default: nil # The javasdk version must be already be installed using ibm-installmgr cookbook. If none is specified the embedded default is used. [See more](http://www.ibm.com/support/knowledgecenter/SS7JFU_8.5.5/com.ibm.websphere.express.doc/ae/rxml_managesdk.html)
* `security_attributes`, [Hash, nil], default: nil # A hash of security attributes to apply to the dmgr
* `manage_user`, [TrueClass, FalseClass], default: true. Should the resource create the user for you. Set to false if you have created the run_user previously in your recipe, outside of this resource.
* `sensitive_exec`, [TrueClass, FalseClass], default: true # set to false for debug purposes to show stdout in Chef logs.
* `timeout`, [Integer, nil], default: nil, The timeout (in seconds) used for starting the Deployment Manager
* `dmgr_svc_timeout`, [Integer, nil], default: 300, The timeout (in seconds) used for starting the Systemd Deployment Manager service

#### Actions

* `:create` - Creates the dmgr
* `:start` - Starts the dmgr.
* `:stop` - Stops the dmgr.
* `:sync_all` - Syncs all federated nodes in the dmgr without restarting the appservers

### websphere_profiles

This resource currently only allows for appserver or custom profiles. It's best to use custom resources everywhere, and add appservers to it later with the `websphere_app_server` or `websphere_cluster_member` resources.

#### Example

```ruby
websphere_profile 'AppProfile1' do
  admin_user 'admin'
  admin_password 'admin'
  dmgr_host 'localhost'
  action [:create, :federate, :start]
end
```

#### Parameters

* `label`, String, name_property: true, just a label to make the execute resource name unique
* `websphere_root`, String, default: '/opt/IBM/WebSphere/AppServer', The WAS install root
* `bin_dir`, String, default: lazy { "#{websphere_root}/bin" } The path to the was root bin directory. No need to change this.
* `admin_user`, [String, nil], default: nil, The dmgr user.
* `admin_password`, [String, nil], default: nil, The dmgr password.  This should only be known to people who know the server is managed with chef.
* `dmgr_host`, String, default: 'localhost', The dmgr host to federate to.
* `dmgr_port`, [String, nil], default: nil, The dmgr port to federate to.
* `profile_name`, String, name_property: true
* `node_name`, String, default: lazy { "#{profile_name}_node" }
* `server_name`, String, default: lazy { "#{profile_name}_server" }
* `cell_name`, [String, nil], default: nil
* `java_sdk`, [String, nil], default: nil # The javasdk version must be already be installed using ibm-installmgr cookbook. If none is specified the embedded default is used. [See more](http://www.ibm.com/support/knowledgecenter/SS7JFU_8.5.5/com.ibm.websphere.express.doc/ae/rxml_managesdk.html)
* `profile_type`, String, default: 'custom', Can be either appserver or custom, but easier to just use custom all the time and add jvms with websphere_app_server or websphere_cluster_member.
* `attributes`, [Hash, nil], default: nil, These are custom attributes for the server when using the appserver type profile, such as monitoringPolicy. They are only set when the node is federated.
* `manage_user`, [TrueClass, FalseClass], default: true. Should the resource create the user for you. Set to false if you have created the run_user previously in your recipe, outside of this resource.
* `manage_service`, [TrueClass, FalseClass], default: true. Should the resource create the init service for you. Set to false if you have created the service previously in your recipe, outside of this resource.
* `sensitive_exec`, [TrueClass, FalseClass], default: true # set to false for debug purposes to show stdout in chef logs.
* `timeout`, [Integer, nil], default: nil, The timeout (in seconds) used for starting the node
* `dmgr_svc_timeout`, [Integer, nil], default: 300, The timeout (in seconds) used for starting the Systemd Deployment Manager service
* `node_svc_timeout`, [Integer, nil], default: 180, The timeout (in seconds) used for starting the Systemd Node agent service

#### Actions

* `:create` - Creates the profile
* `:federate` - Adds the node to a deployment manager for federation/management.
* `:start` - Starts the node.
* `:stop` - Stops the node.
* `:delete` - deletes node and profile and any servers on it. You need to delete any clusters before you can delete the node.

### websphere_app_server

Adds a new jvm to an existing custom profile

#### Example

```ruby
websphere_app_server 'app_server1' do
  node_name 'Custom_node'
  admin_user 'admin'
  admin_password 'admin'
  attributes ({
    'processDefinitions' => {
      'monitoringPolicy' => {
        'newvalue' => "[['maximumStartupAttempts', '2'], ['pingInterval', '222'], ['autoRestart', 'true'], ['nodeRestartState', 'PREVIOUS']]"
      },
      'jvmEntries' => {
        'newvalue' => "[['debugMode', 'true'], ['systemProperties',[[['name','my_custom_websphere_variable'],['value','testing2']]]]]"
      }
    }
  })
  action [:create, :start]
end
```

#### Parameters

* `websphere_root`, String, default: '/opt/IBM/WebSphere/AppServer', The was install root.
* `bin_dir`, String, default: lazy { "#{websphere_root}/bin" } The path to the was root bin directory. No need to change this.
* `admin_user`, [String, nil], default: nil, The dmgr user.
* `admin_password`, [String, nil], default: nil, The dmgr password.  This should only be known to people who know the server is managed with chef.
* `dmgr_host`, String, default: 'localhost', The dmgr host.
* `dmgr_port`, [String, nil], default: nil, The dmgr port leave as nil for default 8879.
* `server_name`, String, name_property: true, The new clustered server name
* `server_node`, [String, nil], default: nil, required: true, The node to create the server on.
* `attributes`, [Hash, nil], default: nil, A hash of attributes to apply to the server, eg. monitoringPolicy.
* `sensitive_exec`, [TrueClass, FalseClass], default: true # set to false for debug purposes to show stdout in chef logs.

#### Actions

* `:create` - Creates the server on node
* `:delete` - Deletes the server
* `:start` - Starts the server

### websphere_cluster

Creates an empty cluster to add app servers to.

#### Example

```ruby
websphere_cluster 'MyCluster' do
  admin_user 'admin'
  admin_password 'admin'
  dmgr_host 'localhost'
  action [:create]
end
```

#### Parameters

* `websphere_root`, String, default: '/opt/IBM/WebSphere/AppServer', The was install root
* `bin_dir`, String, default: lazy { "#{websphere_root}/bin" } The path to the was root bin directory. No need to change this.
* `admin_user`, [String, nil], default: nil, The dmgr user.
* `admin_password`, [String, nil], default: nil, The dmgr password.  This should only be known to people who know the server is managed with chef.
* `dmgr_host`, String, default: 'localhost', The dmgr host to federate to.
* `dmgr_port`, [String, nil], default: nil, The dmgr port to federate to.
* `cluster_name`, String, name_property: true
* `prefer_local`, [TrueClass, FalseClass], default: true
* `sensitive_exec`, [TrueClass, FalseClass], default: true # set to false for debug purposes to show stdout in chef logs.

#### Actions

* `:create` - Creates the cluster
* `:delete` - Deletes the cluster
* `:start` - Starts the servers on the cluster.
* `:ripple_start` - Ripple starts servers on the cluster.

### websphere_cluster_member

Creates and adds an application server to a cluster

#### Example

```ruby
websphere_cluster_member 'ClusterServer1' do
  cluster_name 'MyCluster'
  server_node 'CustomProfile1_Node'
  admin_user 'admin'
  admin_password 'admin'
  member_weight 15
  attributes ({
    'processDefinitions' => {
      'monitoringPolicy' => {
        'newvalue' => "[['maximumStartupAttempts', '2'], ['pingInterval', '222'], ['autoRestart', 'true'], ['nodeRestartState', 'PREVIOUS']]"
      },
      'jvmEntries' => {
        'newvalue' => "[['debugMode', 'true'], ['systemProperties',[[['name','my_custom_websphere_variable'],['value','testing2']]]]]"
      }
    }
  })
  action [:create, :start]
end
```

#### Parameters

* `websphere_root`, String, default: '/opt/IBM/WebSphere/AppServer', The was install root.
* `bin_dir`, String, default: lazy { "#{websphere_root}/bin" } The path to the was root bin directory. No need to change this.
* `admin_user`, [String, nil], default: nil, The dmgr user.
* `admin_password`, [String, nil], default: nil, The dmgr password.  This should only be known to people who know the server is managed with chef.
* `dmgr_host`, String, default: 'localhost', The dmgr host.
* `dmgr_port`, [String, nil], default: nil, The dmgr port leave as nil for default 8879.
* `cluster_name`, String
* `prefer_local`, [TrueClass, FalseClass], default: true
* `server_name`, String, name_property: true, The new clustered server name
* `server_node`, [String, nil], default: nil, required: true, The node to create the server on.
* `member_weight`, Fixnum, default: 2 # number from 0-100. Higher weight gets more traffic
* `session_replication`, [TrueClass, FalseClass], default: false
* `resources_scope`, String, default: 'both', regex: /^(both|server|cluster)$/
* `attributes`, [Hash, nil], default: nil, A hash of attributes to apply to the server, eg. monitoringPolicy.
* `sensitive_exec`, [TrueClass, FalseClass], default: true # set to false for debug purposes to show stdout in chef logs.

#### Actions

* `:create` - Creates the server and adds to cluster
* `:delete` - Deletes the server
* `:start` - Starts the server

### websphere_app

Deploys an application to a server or cluster

#### Example

```ruby
websphere_app 'sample_app_cluster' do
  app_file '/tmp/sample.war'
  cluster_name 'MyCluster'
  admin_user 'admin'
  admin_password 'admin'
  action [:deploy_to_cluster, :start, :stop]
end
```

#### Parameters

* `websphere_root`, String, default: '/opt/IBM/WebSphere/AppServer', The was install root
* `bin_dir`, String, default: lazy { "#{websphere_root}/bin" } The path to the was root bin directory. No need to change this.
* `admin_user`, [String, nil], default: nil, The dmgr user.
* `admin_password`, [String, nil], default: nil, The dmgr password.
* `dmgr_host`, String, default: 'localhost', The dmgr host to federate to.
* `dmgr_port`, [String, nil], default: nil, The dmgr port to federate to.
* `app_name`, String, name_property: true
* `app_file`, [String, nil], default: nil, required: true
* `server_name`, [String, nil], default: nil
* `node_name`, [String, nil], default: nil
* `cluster_name`, [String, nil], default: nil
* `context_root`, [String, nil], default: nil
* `sensitive_exec`, [TrueClass, FalseClass], default: true # set to false for debug purposes to show stdout in chef logs.

#### Actions

* `:deploy_to_cluster` - Deploys app to a cluster
* `:deploy_to_server` - Deploys app to a server
* `:start` - Starts the app
* `:stop` - Stops the app

### websphere_ihs

Deploys an ihs webserver to an existing profile/node.  It currently only supports `managed` ihs webservers.

#### Example

```ruby
websphere_ihs 'MyWebserver1' do
  config_type 'local_distributed'
  profile_name 'AppProfile1'
  node_name 'AppProfile1_node'
  map_to_apps true
  ihs_port '80'
  admin_user 'admin'
  admin_password 'admin'
  action [:create, :start]
end
```

#### Parameters

* `websphere_root`, String, default: '/opt/IBM/WebSphere/AppServer', The was install root
* `bin_dir`, String, default: lazy { "#{websphere_root}/bin" } The path to the was root bin directory. No need to change this.
* `admin_user`, [String, nil], default: nil, The dmgr user.
* `admin_password`, [String, nil], default: nil, The dmgr password.
* `dmgr_host`, String, default: 'localhost', The dmgr host to federate to.
* `dmgr_port`, [String, nil], default: nil, The dmgr port to federate to.
* `webserver_name`, String, name_property: true
* `plg_root`, String, default: '/opt/IBM/WebSphere/Plugins'
* `wct_root`, String, default: '/opt/IBM/WebSphere/Toolbox'
* `wct_tool_root`, String, default: lazy { "#{wct_root}/WCT" }
* `config_type`, String, default: 'local_distributed', Can only be local_distributed or remote.
* `profile_name`, String, required: true
* `node_name`, String, required: true
* `map_to_apps`, [TrueClass, FalseClass], default: false
* `ihs_hostname`, [String],  default: lazy { node.name }
* `ihs_port`, String, default: '80'
* `ihs_install_root`, default: '/opt/IBM/HTTPServer'
* `ihs_config_file`, String, default: lazy { "#{ihs_install_root}/conf/httpd.conf" }
* `runas_user`, String, default: 'root'
* `runas_group`, String, default: 'root'
* `enable_ihs_admin_server`, [TrueClass, FalseClass], default: false
* `ihs_admin_user`, [String, nil], default: nil
* `ihs_admin_password`, [String, nil], default: nil
* `ihs_admin_port`, String, default: '8008'
* `sensitive_exec`, [TrueClass, FalseClass], default: true # set to false for debug purposes to show stdout in chef logs.

#### Actions

* `:create` - Creates a web definition and deploys a webserver
* `:delete` - deletes the web definition
* `:start` - starts the webserver
* `:stop` - stops the webserver

### ibm_cert

Used to manage ibm java certificates.  Create keystore, create certificates and add to keystore, add existing certs to keystore. extract cert from keystore.

This can be used on any server with ikeycmd tool.

#### Example

```ruby
ibm_cert 'mydomain.com' do
  dn 'CN=mydomain.com,O=MyOrg,C=UK'
  kdb '/root/keystores/keystore.kdb'
  kdb_password 'dummy'
  algorithm 'SHA256WithRSA'
  size '2048'
  expire '10'
  owned_by 'jim'
  action [:create, :extract]
end

ibm_cert 'myotherdomain.com' do
  kdb '/root/keystores/keystore.kdb'
  kdb_password 'dummy'
  add_cert '/root/keystores/cert_to_import.cer'
  action :add
end
```

#### Parameters

* `label`, String, name_property: true
* `dn`, String, required: true eg "CN=mydomain.com,O=MyOrg,C=UK"
* `kdb`, String, required: true, regex: Path to keystore file. Will create if it doesn't exist. File must end in .kdb or .p12
* `kdb_password`, String, required: true
* `algorithm`, String, default: 'SHA256WithRSA'
* `size`, String, default: '2048', Can be 2048, 1024 or 512
* `expire`, String, default: '3600', required: true. The expiry for the certificate being created
* `expire_kdb`, String, default: '7300', required: false. The expiry time (in days) for the key database
* `extract_to`, String, Used by the extract action only. Extracts in ASCII to a file <label>.cer next to the keystore file location by default. #{label}.cer" } .
* `add_cert`, [String, nil], default: nil  path to certificate to add to keystore, only used in add.
* `default_cert`, String, default: 'no'. Can be yes or no
* `ikeycmd`, String, default: /opt/IBM/WebSphere/AppServer/java/jre/bin/ikeycmd path to ikeycmd tool.
* `owned_by`, String, default: 'root'
* `sensitive_exec`, [TrueClass, FalseClass], default: true # set to false for debug purposes to show stdout in chef logs.

#### Actions

* `:create` - Creates a keystore with a default certificate
* `:extract` - extracts a certificate from a keystore
* `:add` - adds an existing certificate to an existing keystore.
* `:import` - import a new certificate to an existing keystore.
* `:update` - update a certificate in an existing keystore with the same label.

### websphere_env

Creates a websphere environment variable

#### Example

```ruby
websphere_env 'TESTING_1234' do
  scope 'Cell=Cell1'
  value '/some/value'
  admin_user 'admin'
  admin_password 'admin'
  action :set
end
```

#### Parameters

* `variable_name`, [String, nil], name_property.
* `scope`, String, required: true. eg Cell=Cell1
* `value`, [String, nil], default: nil

#### Actions

* `:set` - creates the variable and sets the value.
* `:remove` - removes the variable.

### websphere_jms_provider

Creates a custom jms provder in websphere

#### Example

```ruby
websphere_jms_provider 'ActiveMQ' do
  scope 'Cell=Cell1'
  context_factory 'org.apache.activemq.jndi.ActiveMQWASInitialContextFactory'
  url 'nsps://devtlnx0287.stbc2.jstest2.net:15071;nsps://devtlnx0293.stbc2.jstest2.net:15071'
  classpath_jars ['/drivers/nClient-9.7.jar', '/drivers/nJMS-9.7.jar']
  description 'ActiveMQ JMS Provider created by Chef'
  admin_user 'admin'
  admin_password 'admin'
  dmgr_host 'localhost'
  action [:create]
end
```

#### Parameters

* `provider_name`, String, name_property: true. Unique name for the custom provider.
* `scope`, String, required: true. eg Cell=Cell1
* `context_factory`, [String, nil], default: nil
* `url`, [String, nil], default: nil
* `classpath_jars`, [Array, nil], default: nil Full path to each jar driver
* `description`, [String, nil], default: nil
* `sensitive_exec`, [TrueClass, FalseClass], default: true # set to false for debug purposes to show stdout in chef logs.

#### Actions

* `:create` - Creates a jms provider

### websphere_jms_conn_factory

Creates a jms connection factory

#### Example

```ruby
websphere_jms_conn_factory 'My Queue Conection Factory' do
  scope 'Cell=Cell1'
  jndi_name 'jms/MyQueueFactory'
  ext_jndi_name 'MyQueueFactory_external_name'
  jms_provider 'ActiveMQ'
  type 'QUEUE'
  conn_pool_timeout '220'
  admin_user 'admin'
  admin_password 'admin'
  dmgr_host 'localhost'
  action [:create]
end
```

#### Parameters

* `conn_factory_name`, String, name_property: true
* `scope`, String, required: true eg 'Cell=MyCell,Cluster=MyCluster'
* `jndi_name`, String, required: true
* `ext_jndi_name`, String, required: true

CHANGELOG.md
IBM Websphere Cookbook CHANGELOG
========================

v1.2.4
--------------------

- add provides to the LWRPs to allow them to work in Chef 16+
- do not require `dn` as a parameter for `ibm_cert` as this isn't required for the `add`/`extract`/`import` etc. actions
- add GitHub Action CI checks

v1.2.3
--------------------

- cookstyle fixes and unit test updates

v1.2.2
--------------------

- Fix incorrect call for dependent services breaking federation.

v1.2.1
--------------------

- Allow dependent services e.g. node agents to start first.

v1.2.0
--------------------

- Add multiple cluster members in a single wsadmin update, rather than separate calls for each.

v1.1.4
--------------------

- Bug fix for `websphere_base`:`get_cluster_members` function, did not return successful output previously

v1.1.3
--------------------

- Bug fix for `websphere_base`:`enable_as_service` function

v1.1.2
--------------------

- Enhance SystemD operations and make it more generic

v1.1.1
--------------------

- Add Systemd services for Deployment Manager and Node Agent

v1.1.0
--------------------

- Chef 13 deprecation fixes
- Additional unit tests for IBM certificate library
- Add a timeout option for Deployment Manager start

v1.0.13
--------------------

- Update Gem dependencies

v1.0.12
--------------------

- Change Deployment Manager init script to use startManager.sh and stopManager.sh
- Allow use of environment variables with the dmgr profile

v1.0.11
--------------------

- update README docs with WAS ND 7.0

v1.0.10
--------------------

- use pure wsadmin to determine cluster membership

v1.0.9
--------------------

- add ability to specify a template for new cluster member

v1.0.8
--------------------

- Don't add a new cluster template if one exists

v1.0.7
--------------------

- Update to allow setting of attributes against a JMS provider

v1.0.6
--------------------

- Fix fingerprint call and make update more idempotent (and faster)

v1.0.5
--------------------

- add update action for IBM certificate import

v1.0.4
--------------------

- add user parameter to commands so files are not owned by root in contexts where we are running as user

v1.0.3
--------------------

- ensure server members with the same name on different nodes do not clash on cluster member checks
- add group setting to certain commands

v1.0.2
--------------------

- use cookstyle for style check
- fix integration tests(IBMCMSKS provider missing)
- do not try to look at the current java sdk if the attribute is not specified (WAS ND 7 Friendly)
- allow to not create a user and a service when using websphere_profile

v1.0.1
--------------------
- run manageprofiles.sh as was run_user
- add sensitive_exec feature for debugging on all resources
- readme update

v1.0.0
--------------------

- released to supermarket. Travis-ci integration tests.

v0.6.0
--------------------

- new websphere env variable resource.

v0.5.2
--------------------

- fix bug when adding a cert that's already in keystore.

v0.5.1
--------------------

- p12 keystores allowed in ibm_cert resource

v0.5.0
--------------------

- new app server resource

v0.4.1
--------------------

- fix bug in jms regex check if exists.
- jms resource fix to avoid custom resource variable scope conflicts.

v0.4.0
--------------------

- new resources for jms destination and jms connection factory. Integration tests for all jms resources.

v0.3.3
--------------------

- cleaner bugfix for starting webserver after it has just been created to allow time to propagate.

v0.3.2
--------------------

- bugfix for starting webserver after it has just been created to allow time to propagate.

v0.3.1
--------------------

- bugfix for federating and existing node that isn't yet federated.

v0.3.0
--------------------

- added vhost_alias resource
- added another sync_node option that doesn't restart nodeagent or servers
- changed extract_cert default filename extension to .cer

v0.2.0
--------------------

- added websphere_wsadmin resource for convenience
- added jms_provider resource
- added ibm_cert resource

v0.1.0
--------------------

- added sync_node feature
- added save config for some changes.
- added start_all_servers in node
- fixed cluster ripplestart and added cluster startSingleServer and stop
- bug fix for libgcc.i686 conflict version with libgcc

v0.0.6
--------------------

- get_ports function to set server endpoints as attributes.
- change logging from warn to debug to reduce noise

v0.0.5
--------------------

- Change default was root to /opt/IBM/WebSphere/Appserver
- add save to various wsadmin commands.

v0.0.4
--------------------

- minor update to ihs integration tests

v0.0.3
--------------------

- bug fix for when security attributes=nil
- add berksfile.lock ot gitignore
- getports.py bug fix

v0.0.2
--------------------

- added global application security property change


libraries/websphere_ihs.rb
#
# Cookbook:: websphere
# Resource:: websphere_ihs
# Copyright:: 2015-2022 J Sainsburys
# License:: Apache License, Version 2.0

module WebsphereCookbook
  class WebsphereIhs < WebsphereBase
    require_relative 'helpers'
    include WebsphereHelpers

    resource_name :websphere_ihs
    provides :websphere_ihs
    property :webserver_name, String, name_property: true
    property :plg_root, String, default: '/opt/IBM/WebSphere/Plugins'
    property :wct_root, String, default: '/opt/IBM/WebSphere/Toolbox'
    property :wct_tool_root, String, default: lazy { "#{wct_root}/WCT" }
    property :config_type, String, default: 'local_distributed', regex: /^(local_distributed|remote)$/ # currently only tested for local_distributed
    property :profile_name, String, required: false # only valid for local_distributed installations
    property :node_name, String, required: true
    property :profile_path, String, default: lazy { "#{websphere_root}/profiles/#{profile_name}" }
    property :map_to_apps, [true, false], default: false
    property :ihs_hostname, [String], default: lazy { node['hostname'] }
    property :ihs_port, String, default: '80'
    property :ihs_install_root, String, default: '/opt/IBM/HTTPServer'
    property :ihs_config_file, String, default: lazy { "#{ihs_install_root}/conf/httpd.conf" }
    property :install_arch, String, default: '64'
    property :runas_user, String, default: 'root'
    property :runas_group, String, default: 'root'
    property :enable_ihs_admin_server, [true, false], default: false
    property :ihs_admin_user, [String, nil]
    property :ihs_admin_password, [String, nil]
    property :ihs_admin_port, String, default: '8008'

    # pre-requisite: this library only supports IHS webservers on managed nodes. The websphere node must already be created on the ihs node.
    # 1. install required packages for wctcmd tool on centos 6
    # 2. install web definition by running wctcmd pct tool
    # 3. execute equivalient to contents of configure script file
    # 4. ensure module is added to apache and restart

    action :create do
      if platform_family?('rhel')
        pkgs = %w(glibc.i686 glibc libgcc.i686)
        pkgs.each do |pkg|
          package pkg
        end
      end

      create_web_definition unless web_definition_exists?
      run_configure_script unless server_exists?(new_resource.node_name, new_resource.webserver_name)

      template '/etc/init.d/ibm-httpd' do
        mode '0755'
        source 'ihs_init.d.erb'
        cookbook 'websphere'
        variables(
          ihs_root: new_resource.ihs_install_root
        )
      end

      service 'ibm-httpd' do
        action :enable
      end

      save_config
      sync_node_wsadmin(new_resource.node_name)
    end

    action :delete do
      delete_web_definition if web_definition_exists?
      save_config
      # TODO: delete server as well?
    end

    action :start do
      i = 0 # safety timeout break
      until server_exists?(new_resource.node_name, new_resource.webserver_name)
        sleep(5) # ensure webserver has been created first
        i += 1
        break if i >= 8
      end
      start_server(new_resource.node_name, new_resource.webserver_name, [0, 103])
    end

    action :stop do
      stop_server(new_resource.node_name, new_resource.webserver_name)
    end

    action :restart do
      i = 0 # safety timeout break
      until server_exists?(new_resource.node_name, new_resource.webserver_name)
        sleep(5) # ensure webserver has been created first
        i += 1
        break if i >= 8
      end
      stop_server(new_resource.node_name, new_resource.webserver_name)
      start_server(new_resource.node_name, new_resource.webserver_name, [0, 103])
    end

    # need to wrap helper methods in class_eval
    # so they're available in the action.
    action_class.class_eval do
      # checks if a web definition has already been defined on this server.
      # you can only have one per definition location
      def web_definition_exists?
        # create webserver definition
        mycmd = Mixlib::ShellOut.new('./wctcmd.sh -tool pct -listDefinitionLocations', cwd: new_resource.wct_tool_root)
        mycmd.run_command
        if mycmd.error?
          Chef::Log.warn("Error checking if web definition exists: #{mycmd.stderr}")
          return false
        end
        if mycmd.stdout.include?('Path:')
          Chef::Log.warn('A Web definition already exists on this server.')
          return true
        end
        Chef::Log.warn('Web definition does NOT exist.')
        false
      end

      # This method runs configureWebserverDefinition.jacl script passing in paramaters.
      # wsadmin.sh -user WASadminUsername -password WASadminPassword -f ${WAS_HOME}/bin/configureWebserverDefinition.jacl
      #  webserverName webserverType webserverInstallLocation webserverConfigFile webserverPort
      #   mapApplications pluginInstallLocation webserverNodeType webserverNodeName webserverHostName webserverOS adminPort adminUserID adminPassword
      # Details of each attribute:
      # WASadminUsername         : User Name for WAS Adminstration
      # WASadminPassword         : Password  for WAS Adminstration
      # webserverName            : Name of the webserver to be defined
      # webserverType            : Type of the web server. Valid values are: IHS APACHE IPLANET DOMINO IIS HTTPSERVER_ZOS
      # webserverInstallLocation : Location of the websserver install
      # webserverConfigFile      : Webserver configuration file
      # webserverPort            : Listening port of the web server
      # mapApplications          : Specifies whether the existing applications are to be mapped to the web server. Valid values are: MAP_NONE MAP_DEFAULT MAP_ALL
      # pluginInstallLocation    : Location of the plugin install
      # webserverNodeType        : Type of the node holding webserver definition, valid values are: managed unmanaged
      # webserverNodeName        : Node at which web server should be defined
      # webserverHostName        : Host name of the web server Needed only for unmanaged node
      # webserverOS              : Operating system of the web server machine Needed only for unmanaged node Valid values are: windows solaris aix hpux linux os400 os390
      # adminPort                : IHS Remote Administration port Needed only for IHS webserver Defaults to 2001 for os400 Defaults to 8008 for other platforms
      # adminUserID              : IHS Remote Administration userid Needed only for IHS webserver
      # adminPassword            : IHS Remote Administration password Needed only for IHS webserver
      # serviceName              : Service name for the webserver Needed only for IHS webserver
      def run_configure_script
        # run configure script on dmgr

        # /opt/IBM/WebSphere/AppServer/profiles/A1_profile/bin/wsadmin.sh -profileName A1_profile -user idadmin -password 99Pr0blems
        # -f '/opt/IBM/WebSphere/AppServer/bin/configureWebserverDefinition.jacl' Webserver01 IHS '/opt/IBM/HTTPServer'
        #     '/opt/IBM/HTTPServer/servers/Webserver01/conf/httpd-Webserver01.conf' 1143 MAP_ALL
        # '/opt/IBM/WebSphere/Plugins' managed ip-10-1-1-191_Node01 ibm-ihs-managed-centos-6 linux

        configure_cmd = "#{new_resource.profile_path}/bin/wsadmin.sh -profileName #{new_resource.profile_name} "\
          "-user #{new_resource.admin_user} -password #{new_resource.admin_password}"\
          " -f '#{new_resource.bin_dir}/configureWebserverDefinition.jacl' "\
          "#{new_resource.webserver_name} IHS '#{new_resource.ihs_install_root}' '#{new_resource.ihs_config_file}' #{new_resource.ihs_port} MAP_ALL '#{new_resource.plg_root}' "\
          "managed #{new_resource.node_name} #{new_resource.ihs_hostname} linux"

        execute "run webserver #{new_resource.webserver_name} configure script" do
          cwd "#{new_resource.profile_path}/bin"
          command configure_cmd
          sensitive true
          action :run
        end
      end

      def create_web_definition
        # create webserver definition
        cmd = "./wctcmd.sh -tool pct -createDefinition -defLocPathname #{new_resource.plg_root} "\
          "-defLocName #{new_resource.webserver_name} -configType #{new_resource.config_type} "\
          "-enableAdminServerSupport #{new_resource.enable_ihs_admin_server} -enableUserAndPass #{new_resource.enable_ihs_admin_server} "\
          "-mapWebServerToApplications #{new_resource.map_to_apps} -profileName #{new_resource.profile_name} -wasExistingLocation #{new_resource.websphere_root} "\
          "-webServerConfigFile1 #{new_resource.ihs_config_file} -webServerDefinition #{new_resource.webserver_name} "\
          "-webServerPortNumber #{new_resource.ihs_port} -webServerSelected ihs "\
          "-webServerHostName #{new_resource.ihs_hostname} -ihsAdminUnixUserGroup #{new_resource.runas_group} -ihsAdminUnixUserID #{new_resource.runas_user}"
        cmd << " -ihsAdminUserID #{new_resource.ihs_admin_user} -ihsAdminPassword #{new_resource.ihs_admin_password}" if new_resource.enable_ihs_admin_server && new_resource.ihs_admin_user && new_resource.ihs_admin_password

        execute "create webserver definition: #{new_resource.webserver_name}" do
          cwd new_resource.wct_tool_root
          command cmd
          sensitive true
          action :run
        end
      end

      def delete_web_definition
        cmd = "./wctcmd.sh -tool pct -removeDefinitionLocation -defLocPathname #{new_resource.plg_root}"
        execute "delete webserver definition: #{new_resource.webserver_name}" do
          cwd new_resource.wct_tool_root
          command cmd
          action :run
        end
      end
    end
  end
end
